ghc-lib-parser 9.12.2 fails to build with ghc-9.8.4 and ghc-lib-parser-9.10.1 fails to build with ghc-9.6.7

Probably related to this https://gitlab.haskell.org/ghc/ghc/-/merge_requests/10568
(patch here https://gitlab.haskell.org/ghc/ghc/-/commit/e421ec3c8fc9688b3f7ee08c0d1baaf1cb7aa750)
`Linking /root/github/stylish-haskell/dist-newstyle/build/x86_64-linux/ghc-9.8.4/stylish-haskell-0.14.6.0/t/stylish-haskell-tests/build/stylish-haskell-tests/stylish-haskell-tests
/usr/lib/gcc/x86_64-pc-linux-gnu/14/../../../../x86_64-pc-linux-gnu/bin/ld: /usr/lib64/ghc-9.8.4/lib/../lib/x86_64-linux-ghc-9.8.4/rts-1.0.2/libHSrts-1.0.2.a(Globals.o):(.bss+0x0): multiple definition of `ghc_unique_counter64'; /root/.local/state/cabal/store/ghc-9.8.4/ghc-lib-parser-9.12.2.20250320-48b5863d60855e9cdc5daef7f57fb49458d3bd0c527dd8950e22b0a76fe73a30/lib/libHSghc-lib-parser-9.12.2.20250320-48b5863d60855e9cdc5daef7f57fb49458d3bd0c527dd8950e22b0a76fe73a30.a(genSym.o):(.bss+0x0): first defined here`

FYI here's the (failing build) output of building stylish-haskell, main branch, with ghc 9.6.7, building with cabal 3.14.2.0.
I'm on Fedora linux:


```
~/Devel/github.com/haskell/stylish-haskell> cabal build
Configuration is affected by the following files:
- cabal.project
Resolving dependencies...
Build profile: -w ghc-9.6.7 -O1
In order, the following will be built (use -v for more details):
 - Cabal-syntax-3.14.2.0 (lib) (requires build)
 - optparse-applicative-0.18.1.0 (lib) (requires build)
 - Cabal-3.14.2.0 (lib) (requires build)
 - stylish-haskell-0.15.0.0 (lib) (configuration changed)
 - stylish-haskell-0.15.0.0 (test:stylish-haskell-tests) (configuration changed)
 - stylish-haskell-0.15.0.0 (exe:stylish-haskell) (first run)
Starting     optparse-applicative-0.18.1.0 (lib)
Starting     Cabal-syntax-3.14.2.0 (lib)
Building     optparse-applicative-0.18.1.0 (lib)
Building     Cabal-syntax-3.14.2.0 (lib)
Installing   optparse-applicative-0.18.1.0 (lib)
Completed    optparse-applicative-0.18.1.0 (lib)
Installing   Cabal-syntax-3.14.2.0 (lib)
Completed    Cabal-syntax-3.14.2.0 (lib)
Starting     Cabal-3.14.2.0 (lib)
Building     Cabal-3.14.2.0 (lib)
Installing   Cabal-3.14.2.0 (lib)
Completed    Cabal-3.14.2.0 (lib)
Configuring library for stylish-haskell-0.15.0.0...
Preprocessing library for stylish-haskell-0.15.0.0...
Building library for stylish-haskell-0.15.0.0...
[23 of 26] Compiling Language.Haskell.Stylish.Config.Cabal ( lib/Language/Haskell/Stylish/Config/Cabal.hs, dist-newstyle/build/x86_64-linux/ghc-9.6.7/stylish-haskell-0.15.0.0/build/Language/Haskell/Stylish/Config/Cabal.o, dist-newstyle/build/x86_64-linux/ghc-9.6.7/stylish-haskell-0.15.0.0/build/Language/Haskell/Stylish/Config/Cabal.dyn_o ) [Source file changed]
[24 of 26] Compiling Language.Haskell.Stylish.Config ( lib/Language/Haskell/Stylish/Config.hs, dist-newstyle/build/x86_64-linux/ghc-9.6.7/stylish-haskell-0.15.0.0/build/Language/Haskell/Stylish/Config.o, dist-newstyle/build/x86_64-linux/ghc-9.6.7/stylish-haskell-0.15.0.0/build/Language/Haskell/Stylish/Config.dyn_o ) [Language.Haskell.Stylish.Config.Cabal changed]
[25 of 26] Compiling Paths_stylish_haskell ( dist-newstyle/build/x86_64-linux/ghc-9.6.7/stylish-haskell-0.15.0.0/build/autogen/Paths_stylish_haskell.hs, dist-newstyle/build/x86_64-linux/ghc-9.6.7/stylish-haskell-0.15.0.0/build/Paths_stylish_haskell.o, dist-newstyle/build/x86_64-linux/ghc-9.6.7/stylish-haskell-0.15.0.0/build/Paths_stylish_haskell.dyn_o ) [dist-newstyle/build/x86_64-linux/ghc-9.6.7/stylish-haskell-0.15.0.0/build/autogen/cabal_macros.h changed]
[26 of 26] Compiling Language.Haskell.Stylish ( lib/Language/Haskell/Stylish.hs, dist-newstyle/build/x86_64-linux/ghc-9.6.7/stylish-haskell-0.15.0.0/build/Language/Haskell/Stylish.o, dist-newstyle/build/x86_64-linux/ghc-9.6.7/stylish-haskell-0.15.0.0/build/Language/Haskell/Stylish.dyn_o ) [Language.Haskell.Stylish.Config changed]
Configuring test suite 'stylish-haskell-tests' for stylish-haskell-0.15.0.0...
Configuring executable 'stylish-haskell' for stylish-haskell-0.15.0.0...
Preprocessing test suite 'stylish-haskell-tests' for stylish-haskell-0.15.0.0...
Building test suite 'stylish-haskell-tests' for stylish-haskell-0.15.0.0...
Preprocessing executable 'stylish-haskell' for stylish-haskell-0.15.0.0...
Building executable 'stylish-haskell' for stylish-haskell-0.15.0.0...
[1 of 1] Compiling Main             ( src/Main.hs, dist-newstyle/build/x86_64-linux/ghc-9.6.7/stylish-haskell-0.15.0.0/x/stylish-haskell/build/stylish-haskell/stylish-haskell-tmp/Main.o )
[ 3 of 16] Compiling Language.Haskell.Stylish.Tests ( tests/Language/Haskell/Stylish/Tests.hs, dist-newstyle/build/x86_64-linux/ghc-9.6.7/stylish-haskell-0.15.0.0/t/stylish-haskell-tests/build/stylish-haskell-tests/stylish-haskell-tests-tmp/Language/Haskell/Stylish/Tests.o ) [Language.Haskell.Stylish changed]
[15 of 16] Compiling Language.Haskell.Stylish.Config.Tests ( tests/Language/Haskell/Stylish/Config/Tests.hs, dist-newstyle/build/x86_64-linux/ghc-9.6.7/stylish-haskell-0.15.0.0/t/stylish-haskell-tests/build/stylish-haskell-tests/stylish-haskell-tests-tmp/Language/Haskell/Stylish/Config/Tests.o ) [Language.Haskell.Stylish.Config changed]
[2 of 2] Linking dist-newstyle/build/x86_64-linux/ghc-9.6.7/stylish-haskell-0.15.0.0/x/stylish-haskell/build/stylish-haskell/stylish-haskell
[16 of 16] Compiling Main             ( tests/TestSuite.hs, dist-newstyle/build/x86_64-linux/ghc-9.6.7/stylish-haskell-0.15.0.0/t/stylish-haskell-tests/build/stylish-haskell-tests/stylish-haskell-tests-tmp/Main.o ) [Language.Haskell.Stylish.Config.Tests changed]
[17 of 17] Linking dist-newstyle/build/x86_64-linux/ghc-9.6.7/stylish-haskell-0.15.0.0/t/stylish-haskell-tests/build/stylish-haskell-tests/stylish-haskell-tests
/usr/bin/ld.gold: error: /home/jhrcek/.ghcup/ghc/9.6.7/lib/ghc-9.6.7/lib/../lib/x86_64-linux-ghc-9.6.7/rts-1.0.2/libHSrts-1.0.2.a(Globals.o): multiple definition of 'ghc_unique_counter64'
/usr/bin/ld.gold: /home/jhrcek/.cabal/store/ghc-9.6.7/ghc-lib-parser-9.10.1.20250103-6822c1bb167729f4c6ea71be204d04266e852a68038c599260f1664e7c2cb181/lib/libHSghc-lib-parser-9.10.1.20250103-6822c1bb167729f4c6ea71be204d04266e852a68038c599260f1664e7c2cb181.a(genSym.o): previous definition here
/usr/bin/ld.gold: error: /home/jhrcek/.ghcup/ghc/9.6.7/lib/ghc-9.6.7/lib/../lib/x86_64-linux-ghc-9.6.7/rts-1.0.2/libHSrts-1.0.2.a(Globals.o): multiple definition of 'ghc_unique_counter64'
/usr/bin/ld.gold: /home/jhrcek/.cabal/store/ghc-9.6.7/ghc-lib-parser-9.10.1.20250103-6822c1bb167729f4c6ea71be204d04266e852a68038c599260f1664e7c2cb181/lib/libHSghc-lib-parser-9.10.1.20250103-6822c1bb167729f4c6ea71be204d04266e852a68038c599260f1664e7c2cb181.a(genSym.o): previous definition here
collect2: error: ld returned 1 exit status
ghc-9.6.7: `gcc' failed in phase `Linker'. (Exit code: 1)
collect2: error: ld returned 1 exit status
ghc-9.6.7: `gcc' failed in phase `Linker'. (Exit code: 1)
Error: [Cabal-7125]
Failed to build exe:stylish-haskell from stylish-haskell-0.15.0.0.
Failed to build test:stylish-haskell-tests from stylish-haskell-0.15.0.0.
```

--- a/ghc-lib-gen/src/Ghclibgen.hs
+++ b/ghc-lib-gen/src/Ghclibgen.hs
@@ -963,11 +963,15 @@ mangleCSymbols ghcFlavor = do
         . prefixSymbol genSym
         . prefixSymbol initGenSym
           =<< readFile' file
-  when (ghcFlavor <= Ghc9121) $
-    let file = "compiler/cbits/genSym.c"
-    in writeFile file
-       . replace "#if !MIN_VERSION_GLASGOW_HASKELL(9,8,4,0)" "#if !MIN_VERSION_GLASGOW_HASKELL(9,6,7,0)"
-       =<< readFile' file
+  let file = "compiler/cbits/genSym.c"
+      replacement = "#if !MIN_VERSION_GLASGOW_HASKELL(9,6,7,0)\n"
+                        <> "HsWord64 ghc_unique_counter64 = 0;\n"
+                        <> "#elif MIN_VERSION_GLASGOW_HASKELL(9,8,0,0) "
+                        <> "&& !MIN_VERSION_GLASGOW_HASKELL(9,8,4,0)"
+   in writeFile file
+        . replace "#if !MIN_VERSION_GLASGOW_HASKELL(9,8,4,0)" replacement
+        . replace "#if !MIN_VERSION_GLASGOW_HASKELL(9,9,0,0)" replacement
+        =<< readFile' file
   when (ghcFlavor == Ghc984) $
     let file = "compiler/cbits/genSym.c"
      in writeFile file

